packages/pieces/community/invoiceninja/src/lib/actions/create-invoice.ts

```diff
@@ -0,0 +1,212 @@
+      defaultValue: true,
+      required: true,
+      options: {
+        options: [
+          {
+            label: 'Amount',
+            value: true
+          },
+          {
+            label: 'Percentage',
+            value: false
+          }
+        ]
+      }
+    }),
+    public_notes: Property.LongText({
+      displayName: 'Public notes for invoice',
+      description: 'Text that may be visible in the client portal (optional)',
+      required: false,
+    }),
+    private_notes: Property.LongText({
+      displayName: 'Private notes for invoice',
+      description: 'Text not visible for clients (optional)',
+      required: false,
+    }),
+    order_items_json: Property.LongText({
+      displayName: 'Order Items JSON string',
+      description: 'e.g., [{ "quantity":1,"product_key":"product key", "discount": "0" }]',
+      required: true,
+    }),
+    send_email: Property.Checkbox({
+      displayName: 'Send invoice to the client by InvoiceNinja e-mail?',
+      description: 'Should we send the invoice to the client on creation?',
+      defaultValue: false,
+      required: true,
+    }),
+    mark_sent: Property.Checkbox({
+      displayName: 'Mark the invoice as sent?',
+      description: 'Makes the invoice active otherwise remains pending.',
+      defaultValue: false,
+      required: true,
+    }),
+    due_date: Property.DateTime({
+      displayName: 'Invoice due date',
+      description: 'e.g., 2024-01-20',
+      required: false,
+      processors: [],
+      validators: [Validators.datetimeIso],
+    }),
+  },
+
+  async run(context) {
+    const INapiToken = context.auth.access_token;
+    const headers = {
+      'X-Api-Token': INapiToken,
+      'Content-Type': 'application/json',
+    };
+
+    const lineItemsArray = JSON.parse(context.propsValue.order_items_json);
+
+    if (!Array.isArray(lineItemsArray)) {
+      throw new Error('Invalid format for order_items_json. It should be an array of objects.');
+    }
+
+    if (lineItemsArray.length === 0) {
+      throw new Error('The line_items array must not be empty.');
+    }
+
+    const isValidLineItem = lineItemsArray.every(item => (
+      typeof item === 'object' &&
+      'quantity' in item && typeof item.quantity === 'number' &&
+      'product_key' in item && typeof item.product_key === 'string' &&
+      'discount' in item && typeof item.discount === 'string'
+    ));
+
+    if (!isValidLineItem) {
+      throw new Error('Each item in the line_items array must be an object with "quantity" (number), "product_key" (string), and "discount" (string).');
+    }
+
+    const baseUrl = context.auth.base_url.replace(/\/$/, '');
+    let errorMessages = '';
+
+    try {
+      const lineItemsWithDetailsPromises = lineItemsArray.map(async item => {
+        try {
+          const getProductDetailsResponse = await fetch(`${baseUrl}/api/v1/products/?product_key=${item.product_key}`, {
```

Comment:
you can use **httpClient** to make a request instead of fetch. see [here](https://www.activepieces.com/docs/developers/building-pieces/create-action#action-definition).

Reply:
Well, it's better for code readability throughout all pieces.But you can change it later.